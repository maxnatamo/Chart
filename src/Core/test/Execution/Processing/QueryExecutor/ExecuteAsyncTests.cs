namespace Chart.Core.Tests.Execution.Processing.QueryExecutorTests
{
    public class ExecuteAsyncTests
    {
        [Fact]
        public async Task ExecuteAsync_ThrowsNullException_GivenEmptyStringRequest()
        {
            // Arrange
            IQueryExecutor executor =
                Schema
                    .Create(c => c.AddType<StringQuery>("Query"))
                    .MakeExecutable();

            // Act
            Func<Task> act = () => executor.ExecuteAsync(requestQuery: string.Empty);

            // Assert
            await act.Should().ThrowExactlyAsync<ArgumentException>();
        }

        [Fact]
        public async Task ExecuteAsync_ThrowsNullException_GivenNullRequest()
        {
            // Arrange
            IQueryExecutor executor =
                Schema
                    .Create(c => c.AddType<StringQuery>("Query"))
                    .MakeExecutable();

            // Act
            Func<Task> act = () => executor.ExecuteAsync(request: null!);

            // Assert
            await act.Should().ThrowExactlyAsync<ArgumentNullException>();
        }

        [Fact]
        public async Task ExecuteAsync_ReturnsOperationCanceledException_GivenCancellingRequest()
        {
            // Arrange
            // Act
            ExecutionResult result =
                await Schema
                    .Create(c => c.AddType<Query>())
                    .MakeExecutable()
                    .ExecuteAsync("{ name }", Query.CancellationTokenSource.Token);

            // Assert
            result.Should().MatchSnapshot();
        }

        [Fact]
        public async Task ExecuteAsync_ReturnsCorrectStringValue()
        {
            // Arrange
            // Act
            ExecutionResult result =
                await Schema
                    .Create(c => c.AddType<StringQuery>("Query"))
                    .MakeExecutable()
                    .ExecuteAsync("{ name }");

            // Assert
            result.MatchSnapshot();
        }

        private class Query
        {
            public static CancellationTokenSource CancellationTokenSource = new();

            public string GetName()
            {
                Query.CancellationTokenSource.Cancel();
                return "Steve";
            }
        }
    }
}