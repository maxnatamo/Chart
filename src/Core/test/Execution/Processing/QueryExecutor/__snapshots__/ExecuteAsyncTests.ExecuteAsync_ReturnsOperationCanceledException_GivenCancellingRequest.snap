{
  "Data": {
    "name": "Steve"
  },
  "Errors": [
    {
      "Message": "Unhandled exception occured while executing the request.",
      "Locations": [],
      "Path": [],
      "Extensions": {}
    }
  ],
  "Extensions": {}
}
